### ISSUE TITLE
direnv for running guix - njson/aliases cannot be found

### ISSUE THREAD

**Issue by simendsjo** [atlas-engineer/nyxt#3014]
I want to improve my development environment, but I'm having some issues.
By adding a `.envrc`, a Guix development environment can be set up automatically when visiting the directory.
But I'm unable to get it to work as there's some issues with `njson/aliases` not being available. Cannot say I understand why though...

The command I've started looks like the following. Notice that I'm adding njson before nyxt.
```
eval $(guix shell \
    --container \
    --network \
    --search-paths \
    --file=_build/njson/guix.scm \
    --development \
    --file=build-scripts/nyxt.scm \
    glib \
    glib-networking \
    gsettings-desktop-schemas \
    git \
    nss-certs)
```

But trying to load nyxt gives 

```
* (ql:quickload :nyxt/gi-gtk)
To load "nyxt/gi-gtk":
  Load 1 ASDF system:
    nyxt/gi-gtk
; Loading "nyxt/gi-gtk"
....To load "njson/aliases":
  Load 1 ASDF system:
    njson
  Install 1 Quicklisp release:
    njson
; Loading "njson/aliases"

debugger invoked on a ASDF/FIND-COMPONENT:MISSING-COMPONENT in thread
#<THREAD tid=8440 "main thread" RUNNING {100AED01B3}>:
  Component "njson/aliases" not found
```

But I run `(ql:quickload :njson)`, which works, and then `njson:

```
* (ql:quickload :njson)
To load "njson":
  Load 1 ASDF system:
    njson
; Loading "njson"
.........
(:NJSON)
* #'njson/aliases:@
#<FUNCTION NJSON/ALIASES:@>
```

Not sure what's going on here. Why cannot njson/aliases be found?

**Comment by aartaka** [comment]
A more direct fix would be to replace `njson/aliases` with a mere `njson`—there was a major version release that integrated the alias system into the main `njson` one :)

**Comment by simendsjo** [comment]
Thanks!

> @simendsjo long story short (and without going into too many details), you're committing a sin when you combine `quicklisp` and `guix`. Can you see why?

I thought Guix would play well together with quicklisp, registering the projects locally so quicklisp would load those rather than from quicklisp.

>  If you're using SLY on Emacs, add the following to your config:

Thanks, very elegant!

> There might be valuable information [here](https://github.com/atlas-engineer/nyxt/blob/master/documents/README.org). If you feel that this document is not good enough, let's improve it together.

For some reason, I'm still not getting this to work.

When starting sly, `cffi:*foreign-library-directories*` is `(#P"/gnu/store/pc3bkj9szmnprlrvach46a968fvs5w2z-profile/lib/")` as I already have a solution for this a bit different from the documentation. I tried using my `~/.guix-profile`, but I get the same result.

I do `(asdf:load-system :nyxt/gtk-application)` and `(nyxt:start)` as stated in the documentation.

![image](https://github.com/atlas-engineer/nyxt/assets/351519/0196168c-151e-4da8-a509-07794534cc8d)

**Comment by aartaka** [comment]
> Thanks!
> 
> > @simendsjo long story short (and without going into too many details), you're committing a sin when you combine `quicklisp` and `guix`. Can you see why?
> 
> I thought Guix would play well together with quicklisp, registering the projects locally so quicklisp would load those rather than from quicklisp.

It kinda does—it still overrides the projects for local ones just fine. The "sin" André refers to here is rather the fact that your environment becomes less reproducible and/or controllable when you start depending on Quicklisp.

**Comment by simendsjo** [comment]
> If you're using SLY on Emacs, add the following to your config:

That's pretty much the same values as the defaults. Tried this exact version, but got the same error.

> @simendsjo after you add the following to your Emacs config, you need to start the SLY REPL via M-- M-x sly RET nyxt-sbcl RET. Notice the negative prefix argument in the beginning. Does that help?

I found out by reading sly's source. I really hate prefix arguments... When adding the nyxt specific implementation, the default value used when it's nil isn't available, so the nyxt one was used. My config looks like the following now

```lisp
;; Custom sly setup for hacking Nyxt on Guix.
(after! sly
  (when-let (file "~/code/nyxt/build-scripts/nyxt-guix.el")
  (load file)
  ;; When sly-lisp-implementations gets a value, the default
  ;; sbcl choice is gone, so we gave to add it ourselves.
  (unless sly-lisp-implementations
    (setq sly-lisp-implementations '((sbcl ("sbcl")))))
  (cl-pushnew '(nyxt (lambda () (nyxt-make-guix-cl-for-nyxt "~/code/nyxt")))
              sly-lisp-implementations)
  ;; TMP: Test to see if some of the other settings have an effect
  (cl-pushnew '(nyxt-from-issue (lambda () (nyxt-make-guix-cl-for-nyxt "~/code/nyxt"
                                                                  :force nil
                                                                  :cl-implementation "sbcl"
                                                                  :cl-system "nyxt/gi-gtk"
                                                                  :preserve '("GDK_SCALE" "GDK_DPI_SCALE")
                                                                  :no-grafts t
                                                                  :ad-hoc '("emacs" "xdg-utils" "git"))))
              sly-lisp-implementations)))
```

> It kinda does—it still overrides the projects for local ones just fine. The "sin" André refers to here is rather the fact that your environment becomes less reproducible and/or controllable when you start depending on Quicklisp.

Yeah, I guess it's easy to suddenly load something from outside guix when doing `ql:quickload`. Good tip.

**Comment by simendsjo** [comment]
> @simendsjo it seems that your nyxt git repository lives at ~/code/nyxt. You need to move it to ~/common-lisp/nyxt, from where ASDF will be able to find it.

I don't think that is the issue as load-system works fine. Tried cloning to ~/common-lisp/nyxt, but I get the same issue.

> I think this might be another source of issues. That variable is nil for me.

Tried setting it to `nil` together with the `~/common-lisp/nyxt` location, but it still breaks exactly the same way:

```
CL-USER> (setf cffi:*foreign-library-directories* nil)
NIL
CL-USER> (asdf:load-system :nyxt/gtk-application)
T
CL-USER> (nyxt:start)
Nyxt version 3.0.0-92-g2d108d24
<INFO> [22:21:12] Source location: #P"/home/simendsjo/common-lisp/nyxt/"
<INFO> [22:21:12] Loading Lisp file #P"/home/simendsjo/.config/nyxt/auto-config.3.lisp".
<INFO> [22:21:12] Loading Lisp file #P"/home/simendsjo/.config/nyxt/config.lisp".
<INFO> [22:21:12] Nyxt-native debugging enabled.
<INFO> [22:21:13] Listening to socket: #P"/run/user/1000/nyxt/nyxt.socket"
<INFO> [22:21:13] GTK extensions directory: #P"/home/simendsjo/common-lisp/nyxt/libraries/web-extensions/"
```

> Maybe the documentation isn't correct, but what you should do is:
> (asdf:load-system :nyxt/gi-gtk)
> (nyxt:start)

The documentation says :nyxt/gtk-application, but both works, and both gives the same result.

> It seems that you're really motivated, so I can offer you some help if you'd be willing to share your screen over Jitsi.

Thanks! We're in the same timezone, so it's late for us both. Let's get in touch tomorrow and find a timeslot.

**Comment by simendsjo** [comment]
> @simendsjo what's the output of `guix describe`? I'm wondering when was the last time you run `guix pull`.

It's just a few days old

```
$ guix describe
Generation 133	Jun 20 2023 08:59:58	(current)
  guix d884fc9
    repository URL: https://git.savannah.gnu.org/git/guix.git
    branch: master
    commit: d884fc9e2efecfba09af4694f5a13ad7fc6f704f
  flat b62ba32
    repository URL: https://github.com/flatwhatson/guix-channel.git
    branch: master
    commit: b62ba3214ed0f781e2d6015044ae8a4a1bd5c7d7
  nonguix bf13c6b
    repository URL: https://gitlab.com/nonguix/nonguix
    branch: master
    commit: bf13c6b5f32980582e8c19b509c6825e86b953dd
```

**Comment by simendsjo** [comment]
Short summary after a video call with André:

While we were unable to locate the root cause, André found a workaround by running `guix shell` directly rather than using the `nyxt-make-guix-cl-for-nyxt` function:

```lisp
(setf sly-lisp-implementations
      `((nyxt ("guix" "shell" "-D" "-f" ,(expand-file-name "~/common-lisp/nyxt/build-scripts/nyxt.scm") "--" "sbcl"))))
```

Nyxt runs fine from the guix repository, when building using `guix package --install-from-file`, and when running the lisp image created by `nyxt-make-guix-cl-for-nyxt`. It only gives the error when started from emacs. We also tried `emacs -Q`, `M-x package-install` `sly`, and then running `nyxt-make-guix-cl-for-nyxt`, but that also failed with the same error.

So there's something wrong with my setup, but we don't know what.

Thanks again for the great help André!

**Comment by aartaka** [comment]
> With respect to the original topic of this issue: @aartaka can we close this or is there anything left to do with respect to `njson`?

No, it's all covered, I believe. NJSON was bumped to version 1, removing all the deprecated code (including `njson/aliases` system). So yeah, it's a matter of updating NJSON.

**Comment by aartaka** [comment]
> When starting sly, cffi:*foreign-library-directories* is (#P"/gnu/store/pc3bkj9szmnprlrvach46a968fvs5w2z-profile/lib/") as I already have a solution for this a bit different from the documentation. I tried using my ~/.guix-profile, but I get the same result.

@simendsjo, you can try looking up the exact lib directory WebKit is installed to, and add it to `cffi:*foreign-library-directories*` before loading Nyxt. Might be worth reinstalling WebKit to see where it ends up, too.

But the error is extremely weird, never seen anything like that 😓
